File: 659-split-array-into-consecutive-subsequences/659-split-array-into-consecutive-subsequences.py
class Solution:
    def isPossible(self, nums: List[int]) -> bool:
        
        fm = {}
        hm = {}
        
        for num in nums:
            fm[num] = fm.get(num, 0) + 1
        
        for x in nums:
            if fm[x] == 0:
                continue
            elif (x in hm) and (hm[x] > 0):
                hm[x] -= 1
                hm[x+1] = hm.get(x+1, 0) + 1
                fm[x] -= 1
            elif (x in fm and fm[x]>0) and (x+1 in fm and fm[x+1]>0) and (x+2 in fm and fm[x+2]>0):
                fm[x] -= 1
                fm[x+1] -= 1
                fm[x+2] -= 1
                hm[x+3] = hm.get(x+3, 0) + 1
            else:
                return False
        return True
    

File: 659-split-array-into-consecutive-subsequences/659-split-array-into-consecutive-subsequences.cpp
class Solution {
public:
    bool isPossible(vector<int>& nums) {
        
        unordered_map<int, int> fm, hm;
        for(auto x : nums){
            fm[x]++;
        }
        
        for(auto x : nums){
            
            if(fm[x] == 0) continue;
            
            else if(hm.find(x) != hm.end() && hm[x] > 0){
                hm[x]--;
                hm[x+1]++;
                fm[x]--;
            }
            else if(fm[x] > 0 && fm[x+1] > 0 && fm[x+2] > 0){
                fm[x]--;
                fm[x+1]--;
                fm[x+2]--;
                hm[x+3]++;
            }
            else return false;
        }
        
        return true;
    }
};